bibtexparser/middlewares/fieldkeys.py

```diff
@@ -0,0 +1,37 @@
+import logging
+from typing import Collection, Dict, List, Set, Union
+
+from bibtexparser.library import Library
+from bibtexparser.model import Block, Entry, Field
+
+from .middleware import BlockMiddleware
+
+class NormalizeFieldKeys(BlockMiddleware):
+    """Normalize field keys to lowercase.
+
+    In case of conflicts (e.g. both 'author' and 'Author' exist in the same entry),
+    a warning is emitted, and the last value wins.
+
+    Some other middlewares, such as `SeparateCoAuthors`, assume lowercase key names.
+    """
+
+    def __init__(self, allow_inplace_modification: bool = True):
+        super().__init__(allow_inplace_modification=allow_inplace_modification,
+                         allow_parallel_execution=True)
+
+    def transform_entry(self, entry: Entry, library: "Library") -> Union[Block, Collection[Block], None]:
+        seen_normalized_keys: Set[str] = set()
+        new_fields_dict: Dict[str, Field] = {}
+        for field in entry.fields:
+            normalized_key: str = field.key.lower()
+            if normalized_key in seen_normalized_keys:
```

Comment:
Since most entries will not have duplicate keys: is it cheaper to first check if the set is smaller than the list, and only then figure out which entries cause conflict?

Reply:
True.

It's a bit complicated to do that way, though, because our `transform_entry` method modifies each entry in-place as it encounters them. The deepcopy for the `allow_inplace_modification=False` mode takes place inside the middleware infrastructure, so also in that case, we mutate the copy we got as input. Thus we always lose the original unnormalized keys during the loop.

OTOH, I agree that performance is a concern, since nowadays big data is a thing, and one would expect libraries like `bibtexparser` to gracefully scale to at least into the 10k...100k range, so it would be nice to solve this.

We could dump a copy of the original unnormalized keys into another container, which we can then compare against the normalized key set after we are done. But then instead of one `if` switch per entry, we must run an `append` or something. All the extra allocations will cause pressure on the GC. Can we expect that solution to be any faster?

Reply:
I see. It appears that you are saying: I can check cheaply on uniqueness, but to have the error message as below, I would need a copy of fields before changing. 

So we have options:
* A more basic error message with just `entry.key`
* A constructor option to choose between basic and advanced error messages
* Keep things as they are

What is your preference? I could live with all

Reply:
Yes, you nailed it. :)

My preference is almost always to go with maximally helpful error messages. It's much better UX for finding the error if the message explicitly says where and what it is. This is something the Python community generally does outstandingly well, and I want to respect that.

But my other preference is to keep things as simple as reasonably possible, following the Zen of Python. In this particular case, is the better UX worth the clarity and maintenance costs of the extra logic to support that UX?

I suppose the tradeoff depends on how long we expect a typical BibTeX entry to be. If it's just a few one-liners - e.g. author/year/title/journal/volume/pages/doi - a duplicate key error should be easy to spot. However, if the entry includes the full abstract, and a long bunch of extra metadata fields from some online service, well, then...

I'll mull over this overnight, and post a solution tomorrow.

Reply:
I've thought about this now.

I think it makes the most sense to keep this part as-is. The `if` is just one check per iteration. The alternative - if we want a helpful error message - is also one operation per iteration, to collect the original key names into a container. But in most cases, that container will never be read (only written to), causing unnecessary GC pressure.

Having a mode switch, on the other hand, increases the complexity. If squeezing out absolute maximal performance is really important, then having two separate implementations is justified, but otherwise I think it just complicates things.

If performance on large datasets becomes an issue, the simplest solution is to simplify the error message (making it less helpful).

I think that in the average application, the performance bottleneck will be somewhere else anyway, i.e. in what one does with the data `bibtexparser` reads in. For example, I'm running an NLP pipeline, semantic embeddings, and UMAP, which (even using GPU acceleration where available) takes ~2 orders of magnitude longer than the BibTeX parse step.

OTOH, use cases where BibTeX parsing makes up the majority of the runtime cost do also exist, such as a reference database format converter. In such cases maximal performance would be useful.

I suppose this comes down to which category of use cases we expect to be more common. My current bet is on applications with nontrivial processing, but please correct if I'm wrong. :)

Reply:
I'm fine with keeping what you suggested. We can always change if needed. Maybe a small comment in the code could be left?

Reply:
Comment added. I'm not sure how to make it small, though. :)